Improve AOTAutograd tests to do something when inputs don't require grad

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106558

This PR:
- Changes the AOTAutograd tests to also check that the output of the
forward is equal under AOTAutograd and eager-mode PyTorch.
- Adds a "check_gradients" flag to `check_aot_autograd`.
  - If True, then we attempt to compute gradients and check them.
  - If False, then we we just check the outputs are equal
  - If "auto", then we will compute gradients and check them only if
    some input and some output requires grad. This option is useful for
    crossref tests where we don't necessarily have inputs that require
    grad.

1) I need a testing utility to test "AOTAutograd for inference",
   e.g. make_fx + functionalize.
2) I want to run aot_autograd_check in crossref tests for other test
suites (e.g. fbgemm) and not all inputs require grad.

Test Plan:
- existing tests
- new tests to test the degenerate cases

zou3519 committed:
Improve AOTAutograd tests to do something when inputs don't require grad

This PR:
- Changes the AOTAutograd tests to also check that the output of the
forward is equal under AOTAutograd and eager-mode PyTorch.
- Changes the AOTAutograd tests to accept inputs that don't require
grad. We don't check gradients if we are unable to compute gradients
due to the inputs not requiring grad or the output of the operator being
non-differentiable.

The main motivations for this PR are that:
1) I need a testing utility to test "AOTAutograd for inference",
   e.g. make_fx + functionalize.
2) I want to run aot_autograd_check in crossref tests for other test
suites (e.g. fbgemm) and not all inputs require grad.

Test Plan:
- existing tests
- new tests to test the degenerate cases

[ghstack-poisoned]

zou3519 committed:
Update on "Improve AOTAutograd tests to do something when inputs don't require grad"


This PR:
- Changes the AOTAutograd tests to also check that the output of the
forward is equal under AOTAutograd and eager-mode PyTorch.
- Changes the AOTAutograd tests to accept inputs that don't require
grad. We don't check gradients if we are unable to compute gradients
due to the inputs not requiring grad or the output of the operator being
non-differentiable.

The main motivations for this PR are that:
1) I need a testing utility to test "AOTAutograd for inference",
   e.g. make_fx + functionalize.
2) I want to run aot_autograd_check in crossref tests for other test
suites (e.g. fbgemm) and not all inputs require grad.

Test Plan:
- existing tests
- new tests to test the degenerate cases

[ghstack-poisoned]

ezyang commented on test/functorch/test_aotdispatch.py:
```diff
@@ -2796,6 +2796,9 @@ def forward(self, x):
     xfail('sparse.mm', 'reduce'),
+    xfail("nn.functional.max_unpool1d"),  # outputs are different
+    xfail("nn.functional.max_unpool2d"),  # outputs are different
+    xfail("nn.functional.max_unpool3d"),  # outputs are different
```
wuh oh

zou3519 replied:
Indeed!

zou3519 replied:
Uh, the problem is just that the sample inputs we generate for max_unpool cause it to be nondeterministic. I can fix that

zou3519 replied:
Fixed the test to not generate inputs that cause it to be nondeterministic

soulitzer commented on torch/testing/_internal/optests/aot_autograd.py:
```diff
@@ -54,10 +66,25 @@ def func_no_tensors(args):
+    any_tensor_requires_grad = pytree.tree_any_only(torch.Tensor, lambda x: x.requires_grad, args)
+    out = wrapper_set_seed(func_no_tensors, args)
+    any_output_requires_grad = pytree.tree_any_only(torch.Tensor, lambda x: x.requires_grad, out)
+    if not any_tensor_requires_grad or not any_output_requires_grad:
```
Maybe we should have users tell us explicitly that they are trying to test inference only, so that errors won't be hidden if they forget to do so

zou3519 replied:
Sure, I can pass in a `supports_autograd={True, False, "auto"}` argument to make it explicit. For the crossref testing we are doing in D47965247 we want something more automatic to avoid needing to specify what the behavior should be.

zou3519 replied:
Flag added

zou3519 committed:
Update on "Improve AOTAutograd tests to do something when inputs don't require grad"


This PR:
- Changes the AOTAutograd tests to also check that the output of the
forward is equal under AOTAutograd and eager-mode PyTorch.
- Changes the AOTAutograd tests to accept inputs that don't require
grad. We don't check gradients if we are unable to compute gradients
due to the inputs not requiring grad or the output of the operator being
non-differentiable.

The main motivations for this PR are that:
1) I need a testing utility to test "AOTAutograd for inference",
   e.g. make_fx + functionalize.
2) I want to run aot_autograd_check in crossref tests for other test
suites (e.g. fbgemm) and not all inputs require grad.

Test Plan:
- existing tests
- new tests to test the degenerate cases

[ghstack-poisoned]

zou3519 committed:
Update on "Improve AOTAutograd tests to do something when inputs don't require grad"


This PR:
- Changes the AOTAutograd tests to also check that the output of the
forward is equal under AOTAutograd and eager-mode PyTorch.
- Changes the AOTAutograd tests to accept inputs that don't require
grad. We don't check gradients if we are unable to compute gradients
due to the inputs not requiring grad or the output of the operator being
non-differentiable.

The main motivations for this PR are that:
1) I need a testing utility to test "AOTAutograd for inference",
   e.g. make_fx + functionalize.
2) I want to run aot_autograd_check in crossref tests for other test
suites (e.g. fbgemm) and not all inputs require grad.

Test Plan:
- existing tests
- new tests to test the degenerate cases

[ghstack-poisoned]

zou3519 committed:
Update on "Improve AOTAutograd tests to do something when inputs don't require grad"


This PR:
- Changes the AOTAutograd tests to also check that the output of the
forward is equal under AOTAutograd and eager-mode PyTorch.
- Adds a "check_gradients" flag to `check_aot_autograd`.
  - If True, then we attempt to compute gradients and check them.
  - If False, then we we just check the outputs are equal
  - If "auto", then we will compute gradients and check them only if
    some input and some output requires grad. This option is useful for
    crossref tests where we don't necessarily have inputs that require
    grad.

1) I need a testing utility to test "AOTAutograd for inference",
   e.g. make_fx + functionalize.
2) I want to run aot_autograd_check in crossref tests for other test
suites (e.g. fbgemm) and not all inputs require grad.

Test Plan:
- existing tests
- new tests to test the degenerate cases

[ghstack-poisoned]

ezyang commented on test/functorch/test_aotdispatch.py:
```diff
@@ -2869,15 +2871,25 @@ def _test_aot_autograd_helper(self, device, dtype, op, dynamic=False):
         try:
             aot_autograd_check(
-                op.op, t_args, t_kwargs, dynamic,
+                op.op, t_args, t_kwargs, dynamic, True,
```
nit: boolean blindness here, kwargify

zou3519 committed:
Update on "Improve AOTAutograd tests to do something when inputs don't require grad"


This PR:
- Changes the AOTAutograd tests to also check that the output of the
forward is equal under AOTAutograd and eager-mode PyTorch.
- Adds a "check_gradients" flag to `check_aot_autograd`.
  - If True, then we attempt to compute gradients and check them.
  - If False, then we we just check the outputs are equal
  - If "auto", then we will compute gradients and check them only if
    some input and some output requires grad. This option is useful for
    crossref tests where we don't necessarily have inputs that require
    grad.

1) I need a testing utility to test "AOTAutograd for inference",
   e.g. make_fx + functionalize.
2) I want to run aot_autograd_check in crossref tests for other test
suites (e.g. fbgemm) and not all inputs require grad.

Test Plan:
- existing tests
- new tests to test the degenerate cases

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "unrelated failures"
